Add layer selector to map

lewfish opened the pull request:
To test, try switching to the ESRI World Imagery base layer using the Leaflet layer selector which should be visible on the map.

Connects #538

caseypt commented on src/mmw/sass/base/_map.scss:
```diff
@@ -24,6 +24,15 @@
 }
+
+.leaflet-control-layers-toggle{
+  background-image: url('/static/images/layers.png');
```
I don't think we should reference `static` in the path, because it can change. FYI, it is configured here: https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/mmw/settings/base.py#L177.

Instead, we should use a relative path.

caseypt replied:
Consider making a whitespace commit to fix the different indentation patterns in this file.

caseypt commented on src/mmw/mmw/settings/base.py:
```diff
@@ -368,3 +368,16 @@ def get_env_setting(setting):
+
+BASE_LAYERS = {
+    'Default': {
+        'url': 'https://{s}.tiles.mapbox.com/v3/ctaylor.lg2deoc9/{z}/{x}/{y}.png',
```
@mmcfarland Is there an Azavea or Stroud Mapbox account we can use?

mmcfarland replied:
Not to my knowledge on either, but they've expressed appreciation of the basemap we've been using.  We'll have to go through the rigmarole of explaining the options and the potential cost of using the MapBox one, unless they have a free tile set that's similar in style to this. I'll put it on the agenda for next week.

caseypt commented on src/mmw/js/src/core/views.js:
```diff
@@ -154,11 +149,36 @@ var MapView = Marionette.ItemView.extend({
+                    maxZoom: 18
+                });
+                map.addLayer(defaultLayer);
+            });
```
I like this workflow for setting up the layer selector. Nice job.

caseypt commented on src/mmw/js/src/core/views.js:
```diff
@@ -154,11 +149,36 @@ var MapView = Marionette.ItemView.extend({
+                L.control.layers(baseLayers).addTo(map);
+                if (defaultLayerName) {
+                    var defaultLayer = baseLayers[defaultLayerName];
+                    map.addLayer(defaultLayer);
```
Consider abstracting out the `done` and `fail` handlers into helper functions.

lliss commented on src/mmw/apps/home/urls.py:
```diff
@@ -4,7 +4,7 @@
 
 from django.conf.urls import patterns, url
-from apps.home.views import home_page, project, compare
+from apps.home.views import home_page, project, compare, base_layers
```
As per @mmcfarland I'd strongly favor having this come in with the initial page rather than as a separate request.

caseypt commented on src/mmw/js/src/core/views.js:
```diff
@@ -154,11 +149,36 @@ var MapView = Marionette.ItemView.extend({
+                        attribution: layerData.attribution || '',
+                        maxZoom: layerData.maxZoom || 18
+                    });
+                }),
```
Indentation seems off here.

caseypt commented on src/mmw/js/src/core/views.js:
```diff
@@ -2,7 +2,7 @@
 var $ = require('jquery'),
     L = require('leaflet'),
-    _ = require('lodash'),
+    _ = require('underscore'),
```
Curious why this was changed. Every function that's in underscore should have a corollary in lodash.

lliss replied:
Might be related to: https://github.com/WikiWatershed/model-my-watershed/issues/450

lewfish replied:
Yes, I did that because I thought we were switching to underscore.

caseypt replied:
@mmcfarland Any thoughts on this? Seems like we should drop one of the `_`'s.

mmcfarland replied:
Yeah, we should drop one.  I'm a little partial to the expanded `lodash` methods, and if there is a version compatible with backbone, I'd be happy to use that.  Aside from that, they're both solid libraries and I haven't got a strong opinion.  I suspect the size difference between the two is negligible, so I think it comes down to backbone support.  If lodash brings in a ton of dependencies, that would also be a strike against it for me.

caseypt commented on src/mmw/js/src/core/tests.js:
```diff
@@ -181,6 +181,54 @@ describe('Core', function() {
+                assert.equal($layers.length, 0, 'Added layer selector');
+
+                view._leafletMap.remove();
+            });
```
Nice tests.

lliss commented on src/mmw/js/src/core/views.js:
```diff
@@ -154,11 +149,36 @@ var MapView = Marionette.ItemView.extend({
+                        return layerData.default;
+                    });
+
+                L.control.layers(baseLayers).addTo(map);
```
I didn't realize the default functionality was that the popup shows on mouse hover. I'm not fond of that effect.

caseypt replied:
Adding `{ collapsed: false }` should fix it.

rajadain replied:
I don't mind the hover / collapse effect too much, although I should probably test it on a touch device. What bothers me more is how different the layer control looks compared to other icons:

![image](https://cloud.githubusercontent.com/assets/1430060/8573149/14d09264-255e-11e5-947c-5216e014cfd2.png)

lliss replied:
Considering the implications on a touch screen is a good call.

rajadain replied:
Just tried it on the iPad, works well. Tap once to expand, then tap to select option. The expanded form stays open until you tap elsewhere.

caseypt commented:
@mmcfarland I'm into current approach as implemented. However, if we want to add more than just the base layer information, it might make more sense to load it all at once rather than having separate endpoints for different config groups or one that returns everything.

This is sort of a tanget, but I'm into this method because our initial page load time is pretty slow. It's mostly because of the JS vendor bundle, but I favor pushing things to async that aren't essential to help with the page load.

mmcfarland commented:
@caseypt Fair point.  I think I'd prefer there be a hierarchical api structure for these thing then, like `/config/base-layers` so they end up too scattered.

lliss commented:
We do need to reduce the vendor bundle if we can. That's one reason to replace lodash with underscore. For some reason I recall though haven't verified recently that lodash was a larger lib that underscore. Either way, I don't think adding a bit of config here is going to slow down the initial load very much. We're talking a few hundred bytes vs a full round trip HTTP request with the SSL handshake. Can't see how the former could possibly be slower.

caseypt commented:
@lliss True that this wouldn't add much to the page load. 

It might be worth seeing how other projects accomplish this, or if there is a recommended way from Django to get settings to the client.

lliss commented:
I'm noticing that all the landcover SVGs are on the initial page. Might be worth moving those off to a separate file since they won't get loaded until the user accesses a project.

caseypt commented:
@lliss has convinced me that the approach suggested by @mmcfarland is probably the way to go. There seems to be a bunch of low hanging fruit to reduce the page load (minify vendor in staging, remove unused or duplicate libraries, off-load things that aren't needed on the initial page), rather than have this async, which won't help as much.

lliss commented:
All other comments aside, I did test this and it does work. It's a totally workable solution though I think we're all leaning towards a reworked implementation to reduce lag a bit.

rajadain commented:
I'm getting the following errors on testing:

![image](https://cloud.githubusercontent.com/assets/1430060/8572726/12dbad24-255c-11e5-9209-0f08dbb6fabf.png)

Steps to reproduce:

 1. Switch to ESRI World Imagery and draw a shape
 2. In the `/analyze` view with the shape visible, select Default tiles
 3. Select ESRI World Imagery tiles. Tiles do not show up and there are two `Cannot read property 'length' of undefined` errors
 3. Select Default tiles. Tiles render, but there are three `Cannot read property 'parentNode' of null` errors.
 4. Keep switching to produce more errors.

kdeloach commented:
@caseypt 

> It might be worth seeing how other projects accomplish this, or if there is a recommended way from Django to get settings to the client.

Here's how OTM2 does it: https://github.com/OpenTreeMap/otm-core/blob/master/opentreemap/treemap/templates/treemap/settings.js

rajadain commented:
@kdeloach 

Regarding using a templated JavaScript file, will it play well with Browserify which will only load it once? How do we load it again after server variables have been changed?

lewfish commented:
I think I've addressed all the comments. I'm now passing in the settings via home.html, which has a script tag that attaches the settings to `window.clientSettings`. The settings are accessed by the app indirectly via core/settings.js which references `window.clientSettings` when they are available, and uses default settings when they are not (which should only occur when running testem tests).

rajadain commented:
Something seems to be missing. `get_home_page_settings` is undefined. There are also other flake8 errors. See the build's console log for details.

lewfish commented:
The build passed. I'll start using the test.sh script before pushing for now on.

rajadain commented:
FYI I'm still seeing the map switching errors in the `/analyze` view

lewfish commented:
After spending some time trying to fix the bug that @rajadain found, and talking about it with @mmcfarland, we decided to postpone any further effort to fix the bug since we might not be showing the layer selector in the Analyze view. https://github.com/WikiWatershed/model-my-watershed/issues/559

rajadain commented:
It might be best to only show the selector in the `/` view, rather than just hide it in the `/analyze` view, because I'm pretty sure that it also doesn't work in the `/project` view.

lliss commented:
Switching the base map after drawing and moving to analyze mode results in errors.

![screencap](https://cloud.githubusercontent.com/assets/903219/8619300/e7cb98a4-26e3-11e5-8bee-0d7a017dee45.png)

lewfish commented:
@lliss This is what I was referring to two comments up and in #559.

lliss commented:
Ok. Then +1.

caseypt commented on src/mmw/apps/core/templates/base.html:
```diff
@@ -32,6 +32,9 @@
 
     {% block javascript %}
+    <script type="text/javascript">
+     window.clientSettings = {{ client_settings | safe }};
```
Indentation
